File: src/main/java/net/igneo/icv/enchantmentActions/enchantManagers/armor/chestplate/PlanarShiftManager.java
package net.igneo.icv.enchantmentActions.enchantManagers.armor.chestplate;

import net.igneo.icv.client.indicators.EnchantIndicator;
import net.igneo.icv.client.indicators.StasisCooldownIndicator;
import net.igneo.icv.enchantment.EnchantType;
import net.igneo.icv.enchantmentActions.Input;
import net.igneo.icv.enchantmentActions.enchantManagers.armor.ArmorEnchantManager;
import net.igneo.icv.init.ICVUtils;
import net.minecraft.core.Direction;
import net.minecraft.world.entity.Entity;
import net.minecraft.world.entity.LivingEntity;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.phys.AABB;
import net.minecraft.world.phys.Vec3;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

public class PlanarShiftManager extends ArmorEnchantManager {
    public PlanarShiftManager(Player player) {
        super(EnchantType.CHESTPLATE, 300, -10, true, player);
    }
    public Vec3 position;
    private List<Entity> blackList = new ArrayList<>();
    @Override
    public void activate() {
        System.out.println("activating");
        position = player.position();
        active = true;
    }

    @Override
    public void onOffCoolDown(Player player) {

    }

    @Override
    public EnchantIndicator getIndicator() {
        return new StasisCooldownIndicator(this);
    }

    @Override
    public boolean canUse() {
        return !active;
    }

    @Override
    public void dualActivate() {
        resetCoolDown();
        Direction dir = player.getDirection();;
        System.out.println(dir);

        float dist = 1;
        Vec3 shiftRot = switch (dir) {
            case DOWN -> null;
            case UP -> null;
            case NORTH ->new Vec3(0,0,-dist);
            case SOUTH -> new Vec3(0,0,dist);
            case WEST -> new Vec3(-dist,0,0);
            case EAST -> new Vec3(dist,0,0);
        };
        double yaw = Math.atan2(shiftRot.z,shiftRot.x);
        float rot = (float) Math.toDegrees(yaw) - 90;
        System.out.println(shiftRot);
        System.out.println(rot);
        System.out.println(ICVUtils.getFlatInputDirection(rot, Input.flattenInput(enchVar.input),10,0));
        Vec3 scale = new Vec3(5,5,5);
        for (Entity entity : player.level().getEntities(null,new AABB(position.subtract(scale),position.add(scale)))) {
            entity.setPos(entity.position().add(ICVUtils.getFlatInputDirection(rot, Input.flattenInput(enchVar.input),10,0)));
        }
        active = false;
        resetCoolDown();
    }

    @Override
    public boolean isDualUse() {
        return true;
    }
}

File: src/main/java/net/igneo/icv/enchantmentActions/enchantManagers/armor/chestplate/PlanarShiftManager.kt
package net.igneo.icv.enchantmentActions.enchantManagers.armor.chestplate

import net.igneo.icv.client.indicators.EnchantIndicator
import net.igneo.icv.client.indicators.StasisCooldownIndicator
import net.igneo.icv.enchantment.EnchantType
import net.igneo.icv.enchantmentActions.Input.Companion.flattenInput
import net.igneo.icv.enchantmentActions.enchantManagers.armor.ArmorEnchantManager
import net.igneo.icv.init.ICVUtils
import net.minecraft.core.Direction
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.phys.AABB
import net.minecraft.world.phys.Vec3
import kotlin.math.atan2

class PlanarShiftManager(player: Player?) :
    ArmorEnchantManager(EnchantType.CHESTPLATE, 300, -10, true, player) {
    var position: Vec3? = null
    private val blackList: List<Entity> = ArrayList()

    override fun activate() {
        println("activating")
        position = player.position()
        active = true
    }

    override fun onOffCoolDown(player: Player?) {
    }

    override val indicator: EnchantIndicator
        get() = StasisCooldownIndicator(this)

    override fun canUse(): Boolean {
        return !active
    }

    override fun dualActivate() {
        resetCoolDown()
        val dir = player.direction
        println(dir)

        val dist = 1f
        val shiftRot = when (dir) {
            Direction.DOWN -> null
            Direction.UP -> null
            Direction.NORTH -> Vec3(0.0, 0.0, -dist.toDouble())
            Direction.SOUTH -> Vec3(0.0, 0.0, dist.toDouble())
            Direction.WEST -> Vec3(-dist.toDouble(), 0.0, 0.0)
            Direction.EAST -> Vec3(dist.toDouble(), 0.0, 0.0)
        }
        val yaw = atan2(shiftRot!!.z, shiftRot.x)
        val rot = Math.toDegrees(yaw).toFloat() - 90
        println(shiftRot)
        println(rot)
        println(
            ICVUtils.getFlatInputDirection(
                rot, flattenInput(
                    enchVar!!.input
                ), 10f, 0.0
            )
        )
        val scale = Vec3(5.0, 5.0, 5.0)
        for (entity in player.level().getEntities(null, AABB(position!!.subtract(scale), position!!.add(scale)))) {
            entity.setPos(
                entity.position().add(
                    ICVUtils.getFlatInputDirection(
                        rot, flattenInput(
                            enchVar!!.input
                        ), 10f, 0.0
                    )
                )
            )
        }
        active = false
        resetCoolDown()
    }

    override var isDualUse: Boolean
        get() = true
        set(isDualUse) {
            super.isDualUse = isDualUse
        }
}


